pkg/store/storepb/matcher_cache.go

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) The Thanos Authors.
```

Comment:
Should we move this code out of `storepb` package? `storepb` sounds more related to the proto itself but this matcher cache can be more generic